File: Gombang/Programmers/미로 탈출 명령어.cpp
#include <string>
#include <vector>

using namespace std;

// 현재 위치에서 이용 가능한 다음 문자를 구하는 함수.
char getAvailableNextChar(int n, int m, int& currX, int& currY)
{
    // 현재 x좌표가 마지막 좌표가 아니라면
    if (currX < n)
    {
        currX++;
        return 'd';
    }
    else if (currY > 1)
    {
        currY--;
        return 'l';
    }
    else if (currY < m)
    {
        currY++;
        return 'r';
    }

    currX--;
    return 'u';
}

string solution(int n, int m, int x, int y, int r, int c, int k) {
    string answer = "";

    // 최소 이동 거리보다 k값이 작으면 impossible.
    int minDistance = abs(r - x) + abs(c - y);
    if (minDistance > k)
        return "impossible";

    // 이동해야 할 x성분과 y성분의 합이 홀수일 때는 k도 홀수여야 하고,
    // 이동해야 할 x성분과 y성분의 합이 짝수일 때는 k도 짝수여야 한다.
    if ((minDistance + k) % 2 != 0)
        return "impossible";

    int currX = x;
    int currY = y;
    int endX = r;
    int endY = c;

    // d → l → r → u
    while (true)
    {
        // 현재 위치에서 목표 위치까지의 거리를 구했을 때 충분히 움직일 수 있는 위치라면
        if (minDistance != k)
        {
            answer += getAvailableNextChar(n, m, currX, currY);
            k--;
        }
        else // 현재 위치에서 목표 위치까지 딱 맞아 떨어지게 움직일 수 있다면 문자열 마무리 정리.
        {
            for (int i = 0; i < endX - currX; i++)
            {
                answer += "d";
            }

            for (int i = 0; i < currY - endY; i++)
            {
                answer += 'l';
            }

            for (int i = 0; i < endY - currY; i++)
            {
                answer += "r";
            }

            for (int i = 0; i < currX - endX; i++)
            {
                answer += "u";
            }

            break;
        }

        minDistance = abs(r - currX) + abs(c - currY);
    }

    return answer;
}

File: Seori/Programmers/미로 탈출 명령어.py
def solution(n, m, x, y, r, c, k):
    # [1] k가 거리보다 작거나, 홀짝이 다른 경우는 불가능
    distance = abs(c - y) + abs(r - x)
    if k < distance or (distance % 2 != k % 2):
        return "impossible"
    
    # [2] 시작점에서 도착점까지의 경로를 구함
    path = ''
    y_move = r - x
    if y_move > 0:
        path += 'd' * y_move
        y_max = r
    else: 
        path += 'u' * (-y_move)
        y_max = x
    
    x_move = c - y
    if x_move > 0:
        path += 'r' * x_move
        x_min = y
    else: 
        path += 'l' * (-x_move)
        x_min = c
    
    # [3] 시작점과 도착점 중 영점(가장 왼쪽 밑)에 가까운 곳에서 영점까지의 경로를 구함.
    #     사전순 d, l, r, u 순에서 유리한게 왼쪽 밑이기 때문
    k -= distance
    while k:
        if y_max < n: 
            path += 'du'
            y_max += 1
            k -= 2
            continue
        
        if x_min > 1:
            path += 'lr'
            x_min -= 1
            k -= 2
            continue
        break
    
    # [4] 지금까지 경로를 사전순으로 정렬하고, 남은 이동 횟수에 따라서 loop(rl 반복)를 추가함
    answer = ''.join(sorted(path))
    loop = 'rl' * (k // 2)

    r_index = answer.find('r')
    if r_index == -1:
        u_index = answer.find('u')
        if u_index == -1:
            return answer + loop
        return answer[:u_index] + loop + answer[u_index:]
    return answer[:r_index] + loop + answer[r_index:]
